### ISSUE TITLE
Extremely long query time for query that takes no time in psql

### ISSUE THREAD

**Issue by cazgp** [brianc/node-postgres#1601]
Running this query:

```
        SELECT
            ISC.table_name "tableName",
            ISC.column_name "columnName",
            ISC.column_default "columnDefault",
            ISC.is_nullable "isNullable",
            ISC.data_type "dataType",
            ISC.character_maximum_length "maxLength",
            ISC.udt_name "udt",
            K.constraint_name "foreignKeyName",
            E.labels
        FROM information_schema.columns ISC
        JOIN information_schema.tables IST
            ON ISC.table_name = IST.table_name
        LEFT JOIN (
            SELECT
                IRC.constraint_name,
                K.table_name,
                K.column_name
            FROM information_schema.key_column_usage K
            JOIN information_schema.referential_constraints IRC
                ON IRC.constraint_name = K.constraint_name
        ) K
            ON (
                K.table_name = ISC.table_name
                AND K.column_name = ISC.column_name
            )
        LEFT JOIN (
            SELECT
                pg_type.typname AS enum_type,
                json_agg(pg_enum.enumlabel) AS labels
            FROM pg_type
            JOIN pg_enum
                ON pg_enum.enumtypid = pg_type.oid
            GROUP BY pg_type.typname
        ) E
            ON (
                ISC.data_type = 'USER-DEFINED'
                AND ISC.udt_name = E.enum_type
            )
        WHERE
            IST.table_schema = 'public'
        AND
            ISC.table_schema = 'public'
        AND
            IST.table_type = 'BASE TABLE'
```

through `pgcli` has the `EXPLAIN ANALYSE` results of Planning time: 8.085 ms and Execution time: 396.976 ms.

When running the exact query through a node-pg query (either with or without SSL, or with or without a pool). It just hangs for a really long time and then eventually responds after about 40s.

Running `SELECT 1` just before running the query responds immediately so there is definitely an open connection to the database and something is happening.

The problem appears to come from

```
        LEFT JOIN (
            SELECT
                IRC.constraint_name,
                K.table_name,
                K.column_name
            FROM information_schema.key_column_usage K
            JOIN information_schema.referential_constraints IRC
                ON IRC.constraint_name = K.constraint_name
        ) K
            ON (
                K.table_name = ISC.table_name
                AND K.column_name = ISC.column_name
            )
```

but running the SELECT part of the query on its own seems to work just fine.

Running the latest version of node-pg (updated a few minutes ago) 7.4.1.

Any help would be greatly appreciated, thank you.

**Comment by sehrope** [comment]
See if you can create a self contained example that reproduces this so that we can see the code. There shouldn't be any difference between executing the same query via node-postgres or via psql.

It's possible the query you're executing isn't *exactly* the same as the one you're running from psql. The use of constants vs parameters could change the query plan as well.

**Comment by cazgp** [comment]
Thank you for the suggestion. I created the small working example and it worked perfectly which made it even more confusing. After a lot of debugging it turns out that running that query on a completely fresh database meant that postgresql had not cached any information and so, for some reason I don't fully understand, was taking forever. This was solved by extracting the problematic SELECT statement into a CTE and the results returned immediately. If anyone wants to do some extracurricular explaining, that would be appreciated, but am closing this as it's got nothing to do with node-pg.
